src/modules/hip/kernel/spatter.hpp

```diff
@@ -0,0 +1,322 @@
+                                         float *spatterMaskInvPtr,
+                                         uint2 *maskLocArr,
+                                         float3 spatterColor,
+                                         RpptROIPtr roiTensorPtrSrc)
+{
+    int id_x = (hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x) * 8;
+    int id_y = hipBlockIdx_y * hipBlockDim_y + hipThreadIdx_y;
+    int id_z = hipBlockIdx_z * hipBlockDim_z + hipThreadIdx_z;
+
+    if ((id_y >= roiTensorPtrSrc[id_z].xywhROI.roiHeight) || (id_x >= roiTensorPtrSrc[id_z].xywhROI.roiWidth))
+    {
+        return;
+    }
+
+    uint srcIdx = (id_z * srcStridesNCH.x) + ((id_y + roiTensorPtrSrc[id_z].xywhROI.xy.y) * srcStridesNCH.z) + (id_x + roiTensorPtrSrc[id_z].xywhROI.xy.x);
+    uint dstIdx = (id_z * dstStridesNH.x) + (id_y * dstStridesNH.y) + id_x * 3;
+    uint maskIdx = (SPATTER_MAX_WIDTH * (maskLocArr[id_z].y + id_y)) + maskLocArr[id_z].x + id_x;
+
+    d_float8 mask_f8, maskInv_f8;
+    mask_f8 = *(d_float8 *)&spatterMaskPtr[maskIdx];
+    maskInv_f8 = *(d_float8 *)&spatterMaskInvPtr[maskIdx];
+    float4 r_f4 = (float4)(spatterColor.x);
+    float4 g_f4 = (float4)(spatterColor.y);
+    float4 b_f4 = (float4)(spatterColor.z);
+
+    d_float24 src_f24, dst_f24;
+
+    rpp_hip_load24_pln3_and_unpack_to_float24_pln3(srcPtr, srcIdx, srcStridesNCH.y, &src_f24);
+    spatter_hip_compute(srcPtr, &src_f24.x, &dst_f24.x, &mask_f8, &maskInv_f8, &r_f4);
+    spatter_hip_compute(srcPtr, &src_f24.y, &dst_f24.y, &mask_f8, &maskInv_f8, &g_f4);
+    spatter_hip_compute(srcPtr, &src_f24.z, &dst_f24.z, &mask_f8, &maskInv_f8, &b_f4);
+    rpp_hip_pack_float24_pln3_and_store24_pkd3(dstPtr, dstIdx, &dst_f24);
+}
+
+template <typename T>
+RppStatus hip_exec_spatter_tensor(T *srcPtr,
+                                  RpptDescPtr srcDescPtr,
+                                  T *dstPtr,
+                                  RpptDescPtr dstDescPtr,
+                                  uint2 *maskLocArr,
+                                  RpptRGB spatterColor,
+                                  RpptROIPtr roiTensorPtrSrc,
+                                  rpp::Handle& handle)
+{
+    int localThreads_x = 16;
+    int localThreads_y = 16;
+    int localThreads_z = 1;
+    int globalThreads_x = (dstDescPtr->strides.hStride + 7) >> 3;
+    int globalThreads_y = dstDescPtr->h;
+    int globalThreads_z = handle.GetBatchSize();
+
+    float3 spatterColor_f3;
+    if (dstDescPtr->c == 3)
+    {
+        spatterColor_f3 = make_float3((float)spatterColor.B, (float)spatterColor.G, (float)spatterColor.R);
+    }
+    else if (dstDescPtr->c == 1)
+    {
+        float meanGreyVal = ((float)spatterColor.B + (float)spatterColor.G + (float)spatterColor.R) * 0.3333;
+        spatterColor_f3 = make_float3(meanGreyVal, meanGreyVal, meanGreyVal);
+    }
+
+    float *spatterMaskPtr, *spatterMaskInvPtr;
+    Rpp32u maskSize = 2073600 * sizeof(float);
+    hipMalloc(&spatterMaskPtr, maskSize);
```

Comment:
you should not allocate memory here during the launch, the allocation should occur outside of this function and pass a pointer to this hip_exec_spatter_tensor. Please remove line 238-241 from here.

Reply:
I've moved the allocation out. Now pre-allocating temp hip buffers like for other functions.
I presume copying new data into a pre-allocated buffer is okay to have in the call path?